File: design-patterns/ts/observer/general/concrete-subject.ts
// concrete-subject.ts

import Observer from "./observer";
import Subject from "./subject";

class ConcreteSubject implements Subject {
    private state: number = 0;
    private observerList: Observer[] = [];

    getState(): number {
        return this.state;
    }

    setState(state: number): void {
        this.state = state;
        this.notifyObservers();
    }

    public attach(observer: Observer): void {
        this.observerList.push(observer);
    }

    public notifyObservers(): void {
        for (let observer of this.observerList) {
            observer.sendUpdate()
        }
    }
}

export default ConcreteSubject;

File: design-patterns/go/observer/general/concrete_subject.go
// concrete_subject.go

package main

type ConcreteSubject struct {
	state        int
	observerList []IObserver
}

func NewConcreteSubject() (concreteSubject *ConcreteSubject) {
	concreteSubject = &ConcreteSubject{}
	return
}

func (concreteSubject *ConcreteSubject) Attach(observer IObserver) {
	concreteSubject.observerList = append(concreteSubject.observerList, observer)
}

func (concreteSubject *ConcreteSubject) GetState() int {
	return concreteSubject.state
}

func (concreteSubject *ConcreteSubject) NotifyObservers() {
	for _, observer := range concreteSubject.observerList {
		observer.SendUpdate()
	}
}

func (concreteSubject *ConcreteSubject) SetState(state int) {
	concreteSubject.state = state
	concreteSubject.NotifyObservers()
}
